### DISCUSSION TITLE
bashreadline.bt shows only output for new shells

### DISCUSSION THREAD

**Discussion by JLT032** [bpftrace/bpftrace#2101] [Q&A]
learning BPF i noticed `sudo bashreadline.bt` only outputs bash commands for new shell reliably. The existing shell returned one typed command, using history did not return anything, after this one command no other command was shown. creating a new bash shell shows all commands.

my question is if this is expected behavior

**Comment by brendangregg** [comment]
No, that behavior doesn't happen for me. Existing shell:

```
/home/bgregg/Git/bpftrace/tools/bashreadline.bt 
Attaching 2 probes...
Tracing bash commands... Hit Ctrl-C to end.
TIME      PID    COMMAND
11:58:47  6009   history
11:58:52  6009   echo works
```

Note that the first command is missed.

If you have a problem, you'll need to debug it as I can't reproduce.

**Reply by fbs** [reply]
One thing to check is whether the breakpoint has been set correctly, in the shell that doens't work:


``` 
[root@hackbox ~]# gdb -p $$ -batch -ex 'disas readline'
0x00007f3f498b54fc in waitpid () from /lib64/libc.so.6
Dump of assembler code for function readline:
   0x000000000048aea0 <+0>:	int3
   0x000000000048aea1 <+1>:	push   %rsp
```

after unloading the bpf prog

```
[root@hackbox ~]# gdb -p $$ -batch -ex 'disas readline'
0x00007f3f498b54fc in waitpid () from /lib64/libc.so.6
Dump of assembler code for function readline:
   0x000000000048aea0 <+0>:	push   %r12
   0x000000000048aea2 <+2>:	push   %rbp
```

the int3 should be there to trigger the bpfprog

**Reply by JLT032** [reply]
wow, learning a lot here. As a verification, this shell is GNU bash, version 5.1.8(1)-release (x86_64-pc-linux-gnu) on Ubuntu 21.10

**Reply by JLT032** [reply]
I've actually just replicated the situation. 

1. sudo bashreadline.bt
2. sudo gdb -p $$ -batch -ex 'disas readline' 
3. ls
4. ls /
5. ls
6. cd
7. ls

Where 2,3,6,7 are not visible in the output for 1 but 4 is.
It seems there is a way to mask 'stuff' from 'bashreadline.bt'

1. cd
2. cp logs-21-11-18_17-51-47.zip blogs.zip
3. cp logs-21-11-18_17-51-47.zip blogs.zip
4. cp logs-21-11-18_17-51-47.zip blogs.zip

where only 4 is visible

It must be mentioned i do have some sysctl restrictions in place but these only apply to non-privileged users.

**Reply by fbs** [reply]
Do you run 1 and 2 in separate terminals?

If 1 and 2 run in separate terminals(t1 & t2), both existing before 1 is executed than 2 should not show up but 3-7 (ran in t2) should show up. Its weird that 4&5 do but 6 doesn't.

Can you set gdb to break on `readline`, if gdb never hits the breakpoint then it (probably) makes sense that the probe doesn't fire

**Reply by JLT032** [reply]
hey @fbs 

For each example. I start the bpftrace script bashreadline.bt in one terminal then i open another terminal and perform all of the stepts mentioned. I do not start a new terminal session at any time.

These are two separate examples. I found whenever i perform 'cd' the next lines are masked from bashreadline.bt, this is what i document in the second example.

please explain what you mean with "Can you set gdb to break on readline"

Happy 2022 :-)
